File: 61.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return head;
        int length = 0;
        ListNode countLength = head;
        while (countLength != null) {
            countLength = countLength.next;
            length ++;
        }
        k %= length;
        ListNode header = new ListNode(-1, head);
        ListNode slow = head, fast = head;
        for (int i = 0; i < k; ++i) {
            fast = fast.next;
        }
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        fast.next = header.next;
        header.next = slow.next;
        slow.next = null;
        return header.next;
    }
}

File: 61.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) return head;;
        int length = 0;
        for (ListNode * pointer = head; pointer != nullptr; pointer = pointer->next)
            length ++;
        k %= length;
        if (k == 0) return head;
        ListNode header(-1, head);
        ListNode * fast_pointer = head, * slow_pointer = head;
        for (int i = 0; i < k; ++i) {
            fast_pointer = fast_pointer->next;
        }
        while (fast_pointer->next != nullptr) {
            slow_pointer = slow_pointer->next;
            fast_pointer = fast_pointer->next;
        }
        ListNode * new_head = slow_pointer->next;
        slow_pointer->next = nullptr;
        fast_pointer->next = header.next;
        return new_head;
    }
};